depthai_ros_driver/src/conversion.cpp

```diff
@@ -0,0 +1,107 @@
+        case dai::RawImgFrame::Type::RGB888p:
+            cv_size = cv::Size(cols, rows);
+            cv_type = CV_8UC3;
+            break;
+
+        case dai::RawImgFrame::Type::YUV420p:
+        case dai::RawImgFrame::Type::NV12:
+        case dai::RawImgFrame::Type::NV21:
+            cv_size = cv::Size(cols, rows * 3 / 2);
+            cv_type = CV_8UC1;
+            break;
+
+        case dai::RawImgFrame::Type::RAW8:
+        case dai::RawImgFrame::Type::GRAY8:
+            cv_size = cv::Size(cols, rows);
+            cv_type = CV_8UC1;
+            break;
+
+        case dai::RawImgFrame::Type::GRAYF16:
+            cv_size = cv::Size(cols, rows);
+            cv_type = CV_16FC1;
+            break;
+
+        case dai::RawImgFrame::Type::RAW16:
+            cv_size = cv::Size(cols, rows);
+            cv_type = CV_16UC1;
+            break;
+
+        case dai::RawImgFrame::Type::RGBF16F16F16i:
+        case dai::RawImgFrame::Type::BGRF16F16F16i:
+        case dai::RawImgFrame::Type::RGBF16F16F16p:
+        case dai::RawImgFrame::Type::BGRF16F16F16p:
+            cv_size = cv::Size(cols, rows);
+            cv_type = CV_16FC3;
+            break;
+
+        case dai::RawImgFrame::Type::BITSTREAM:
+        default:
+            cv_size = cv::Size(static_cast<int>(size), 1);
+            cv_type = CV_8UC1;
+            break;
+    }
+    return cv::Mat(cv_size, cv_type);
+}
+
+cv::Mat convert_img(const depthai_datatype_msgs::RawImgFrame& input) {
+    // prepare cv_mat
+    cv::Mat mat = prepare_cv_mat_from_dai_type(
+            static_cast<dai::RawImgFrame::Type>(input.fb.type), input.fb.height, input.fb.width, input.data.size());
+
+    // Check if enough data
+    long requiredSize = CV_ELEM_SIZE(mat.type()) * mat.size().area();
+    long actualSize = static_cast<long>(input.data.size());
+    if (actualSize < requiredSize) {
+        ROS_WARN_STREAM("ImgFrame doesn't have enough data to encode specified frame, required "
+                        << requiredSize << ", actual " << actualSize << ". Maybe metadataOnly transfer was made?");
+    } else if (actualSize > requiredSize) {
+        ROS_WARN_STREAM("ImgFrame has too much data to encode specified frame, required " << requiredSize << ", actual "
+                                                                                          << actualSize << ".");
+    }
+    if (input.fb.width <= 0 || input.fb.height <= 0) {
+        ROS_WARN("ImgFrame metadata not valid (width or height = 0)");
+    }
+
+    // actual copy to cv_mat
+    assert(input.data.size() <= static_cast<long>(mat.dataend - mat.datastart));
+    std::memcpy(mat.data, input.data.data(), input.data.size());
+
+    // handle weird case
+    if (static_cast<dai::RawImgFrame::Type>(input.fb.type) == dai::RawImgFrame::Type::NV12) {
+        // weird case when 'video' image is type dai::RawImgFrame::Type::NV12 instead of BGR
+        cv::cvtColor(mat, mat, cv::COLOR_YUV2BGR_NV12);
```

Comment:
We convert the image to BGR_NV12? Or from it?

Reply:
YUV_NV12 to BGR. confusing name.